[AMDGPU] Use table strategy for LowerModuleLDSPass at O0

hjagasiaAMD opened the pull request:
Ensure global variables accessed by only one kernel can stay in kernel scope at O0 by switching to table strategy for AMDGPULowerModuleLDSPass. This to prevent LDS limit from being exceeded for the kernel. At higher Opt levels, additional passes run can acheive this without switching to table strategy.

hjagasiaAMD committed:
[AMDGPU] Use table strategy for LowerModuleLDSPass at O0

Ensure global variables accessed by only one kernel can stay in kernel
scope at O0 by switching to table strategy for AMDGPULowerModuleLDSPass.
This to prevent LDS limit from being exceeded for the kernel. At higher
Opt levels, additional passes run can acheive this without switching to
table strategy.

hjagasiaAMD committed:
Merge branch 'main' into hjagasiaAMD

hjagasiaAMD committed:
[AMDGPU] Use table strategy for LowerModuleLDSPass if LDS limit exceeded

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULowerModuleLDSPass.cpp:
```diff
@@ -648,7 +649,19 @@ class AMDGPULowerModuleLDS {
         } else if (set_is_subset(K.second, HybridModuleRootKernels)) {
-          ModuleScopeVariables.insert(GV);
+          uint64_t LocalMemLimit = 0;
+          for (Function &F : M) {
```
Comment what this is doing

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULowerModuleLDSPass.cpp:
```diff
@@ -648,7 +649,19 @@ class AMDGPULowerModuleLDS {
+          for (Function &F : M) {
+            if (!F.isDeclaration()) {
+              const GCNSubtarget &ST = TM.getSubtarget<GCNSubtarget>(F);
+              LocalMemLimit = ST.getAddressableLocalMemorySize();
```
The limit should really come from the entry point kernel, not just the first function you happen to find

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULowerModuleLDSPass.cpp:
```diff
@@ -648,7 +649,19 @@ class AMDGPULowerModuleLDS {
           KernelAccessVariables.insert(GV);
         } else if (set_is_subset(K.second, HybridModuleRootKernels)) {
-          ModuleScopeVariables.insert(GV);
+          uint64_t LocalMemLimit = 0;
```
The comment at the top of the file claims the hybrid strategy offers precise allocation, so is there just a bug somewhere?

arsenm commented on llvm/test/CodeGen/AMDGPU/lower-module-lds-force-table.ll:
```diff
@@ -0,0 +1,237 @@
+  ret void
+}
+
+define i32 @_Z17HlifCompressBatchILi1ERN7hipcomp25cascaded_compress_wrapperIhmLi128ELi4096EEERN18cooperative_groups12thread_blockEEvRK12CompressArgsOT0_OT1_() {
```
Fix all of these variable names, replace them with more meaningful testcase names

arsenm commented on llvm/test/CodeGen/AMDGPU/lower-module-lds-force-table.ll:
```diff
@@ -0,0 +1,237 @@
+; RUN: llc -mtriple=amdgcn-amd-amdhsa < %s
```
Need to check the output, preferably of the IR pass

hjagasiaAMD committed:
[AMDGPU] Use table strategy for LowerModuleLDSPass at O0

Ensure global variables accessed by only one kernel can stay in kernel
scope at O0 by switching to table strategy for AMDGPULowerModuleLDSPass.
This to prevent LDS limit from being exceeded for the kernel. At higher
Opt levels, additional passes run can acheive this without switching to
table strategy.

JonChesterfield commented:
The different strategies change how variables are accessed but not where they are allocated. This change may improve compile time. It will definitely regress runtime.

If it changes the the reported amount of lds used by any kernel, there is an error in this pass or elsewhere.

arsenm commented on llvm/test/CodeGen/AMDGPU/lower-module-lds-force-table.ll:
```diff
@@ -0,0 +1,264 @@
+; RUN: opt -S -mtriple=amdgcn-- -mcpu=gfx942 -amdgpu-lower-module-lds < %s 2>&1 | FileCheck %s
+; RUN: opt -S -mtriple=amdgcn-- -mcpu=gfx942 -passes=amdgpu-lower-module-lds < %s 2>&1 | FileCheck %s
+
```
This test isn't reduced enough. I previously posted a godbolt link which is smaller than this, and I'm sure it can be shrunk further

JonChesterfield replied:
I see the godbolt IR. In what sense is this believed to change allocated lds size?

JonChesterfield reviewed (changes requested):
The change is noisy enough that I'm not sure what it's trying to do. The goal stated in the commit message can't be met by the change stated in the commit message.

What's the test case that this change decreases the amount of allocated lds for?

Tangentially related, was this report prepared using an LLM?

JonChesterfield commented:
Reproduces (on whatever I had in ~/llvm-install)

```
~/llvm-install/bin/opt -mtriple=amdgcn -mcpu=gfx942 lower-lds.ll -amdgpu-lower-module-lds -S -o lower-lds.baseline.ll

~/llvm-install/bin/opt -mtriple=amdgcn -mcpu=gfx942 lower-lds.ll -amdgpu-lower-module-lds -amdgpu-lower-module-lds-strategy=table -S -o lower-lds.table.ll 
```

The backend allocates static lds based on the amdgpu-lds-size metadata. These numbers are different for the above to paths but shouldn't be.

I suspect we're counting it wrong on the table path, as opposed to the default properly allocating 68k and the table one properly allocating between 11k and 26k, but there's definitely something not working as intended here. Thanks for the reproducer

```
; lower-lds.ll
%RawStorage1 = type { [1056 x i8] }
%RawStorage2 = type { [4 x i8] }
%RawStorage3 = type { [16 x i8] }

@one = addrspace(3) global [1026 x i32] poison
@two = addrspace(3) global [1026 x i32] poison
@three = external addrspace(3) global [2048 x i32]
@four = addrspace(3) global [2050 x i32] poison
@five = addrspace(3) global [16 x i32] poison
@six = external addrspace(3) global %RawStorage1
@seven = addrspace(3) global %RawStorage2 poison
@eight = addrspace(3) global [1026 x i32] poison
@nine = addrspace(3) global [1026 x i32] poison
@ten = external addrspace(3) global [1024 x i32]
@eleven = addrspace(3) global [1026 x i32] poison
@twelve = external addrspace(3) global [16 x i32]
@thirteen = external addrspace(3) global %RawStorage1
@fourteen = external addrspace(3) global [1 x i32]
@fifteen = addrspace(3) global [1026 x i32] poison
@sixteen = addrspace(3) global [1026 x i32] poison
@seventeen = external addrspace(3) global [512 x i32]
@eighteen = addrspace(3) global [514 x i32] poison
@nineteen = external addrspace(3) global [16 x i32]
@twenty = external addrspace(3) global %RawStorage1
@twentyone = external addrspace(3) global [514 x i64]
@twentytwo = external addrspace(3) global [514 x i64]
@twentythree = external addrspace(3) global [256 x i32]
@twentyfour = external addrspace(3) global [258 x i32]
@twentyfive = external addrspace(3) global [16 x i32]
@twentysix = external addrspace(3) global %RawStorage1
@twentyseven = external addrspace(3) global %RawStorage3

define amdgpu_kernel void @EN32_compress_wrapperIhm() {
entry:
  %0 = call i32 @Ihm_one()
  ret void
}

define i32 @Ihm_one() {
entry:
  %0 = call i32 @Ihm_chunk()
  ret i32 %0
}

define i32 @Ihm_chunk() {
entry:
  %0 = call i32 @Ihm_CascadedOpts()
  ret i32 %0
}

define i32 @Ihm_CascadedOpts() {
entry:
  store ptr addrspacecast (ptr addrspace(3) @one to ptr), ptr null, align 8
  store ptr addrspacecast (ptr addrspace(3) @two to ptr), ptr null, align 8
  %add.ptr = getelementptr i32, ptr getelementptr inbounds (i32, ptr addrspacecast (ptr addrspace(3) @five to ptr), i64 1), i64 0
  call void @Ihm_PS1_PT1_PS4_S7()
  %call69 = call i32 @foo(ptr addrspacecast (ptr addrspace(3) @three to ptr), ptr addrspacecast (ptr addrspace(3) @four to ptr))
  ret i32 %call69
}

define void @Ihm_PS1_PT1_PS4_S7() {
entry:
  store ptr addrspacecast (ptr addrspace(3) @six to ptr), ptr null, align 8
  ret void
}

define i32 @foo(ptr %input, ptr %temp_storage) {
entry:
  call void @Itm_PjPS4()
  ret i32 0
}

define void @Itm_PjPS4() {
entry:
  call void @Itm_PS1_Pj()
  ret void
}

define void @Itm_PS1_Pj() {
entry:
  store ptr addrspacecast (ptr addrspace(3) @seven to ptr), ptr null, align 8
  ret void
}

define amdgpu_kernel void @EN32_compress_wrapperItm() {
entry:
  %0 = call i32 @Itm_one()
  ret void
}

define i32 @Itm_one() {
entry:
  %0 = call i32 @Itm_chunk()
  ret i32 %0
}

define i32 @Itm_chunk() {
entry:
  %0 = call i32 @Itm_CascadedOpts()
  ret i32 %0
}

define i32 @Itm_CascadedOpts() {
entry:
  store ptr addrspacecast (ptr addrspace(3) @eight to ptr), ptr null, align 8
  store ptr addrspacecast (ptr addrspace(3) @nine to ptr), ptr null, align 8
  %add.ptr = getelementptr i32, ptr getelementptr inbounds (i32, ptr addrspacecast (ptr addrspace(3) @twelve to ptr), i64 1), i64 0
  call void @Itm_PS1_PT1_PS4_S7()
  %call69 = call i32 @foo(ptr addrspacecast (ptr addrspace(3) @ten to ptr), ptr addrspacecast (ptr addrspace(3) @eleven to ptr))
  ret i32 %call69
}

define void @Itm_PS1_PT1_PS4_S7() {
entry:
  store ptr addrspacecast (ptr addrspace(3) @thirteen to ptr), ptr null, align 8
  ret void
}

define amdgpu_kernel void @EN32_compress_wrapperIjm() {
entry:
  %arrayidx = getelementptr [1 x i32], ptr addrspacecast (ptr addrspace(3) @fourteen to ptr), i64 0, i64 0
  %0 = call i32 @Ijm_one()
  ret void
}

define i32 @Ijm_one() {
entry:
  %0 = call i32 @Ijm_chunk()
  ret i32 %0
}

define i32 @Ijm_chunk() {
entry:
  %0 = call i32 @Ijm_CascadedOpts()
  ret i32 %0
}

define i32 @Ijm_CascadedOpts() {
entry:
  store ptr addrspacecast (ptr addrspace(3) @fifteen to ptr), ptr null, align 8
  store ptr addrspacecast (ptr addrspace(3) @sixteen to ptr), ptr null, align 8
  %add.ptr = getelementptr i32, ptr getelementptr inbounds (i32, ptr addrspacecast (ptr addrspace(3) @nineteen to ptr), i64 1), i64 0
  call void @Ijm_PS1_PT1_PS4_S7()
  %call69 = call i32 @foo(ptr addrspacecast (ptr addrspace(3) @seventeen to ptr), ptr addrspacecast (ptr addrspace(3) @eighteen to ptr))
  ret i32 %call69
}

define void @Ijm_PS1_PT1_PS4_S7() {
entry:
  store ptr addrspacecast (ptr addrspace(3) @twenty to ptr), ptr null, align 8
  ret void
}

define amdgpu_kernel void @EN32_compress_wrapperImm() {
entry:
  %0 = call i32 @Imm_one()
  ret void
}

define i32 @Imm_one() {
entry:
  %0 = call i32 @Imm_chunk()
  ret i32 %0
}

define i32 @Imm_chunk() {
entry:
  %0 = call i32 @Imm_CascadedOpts()
  ret i32 %0
}

define i32 @Imm_CascadedOpts() {
entry:
  store ptr addrspacecast (ptr addrspace(3) @twentyone to ptr), ptr null, align 8
  store ptr addrspacecast (ptr addrspace(3) @twentytwo to ptr), ptr null, align 8
  %add.ptr = getelementptr i32, ptr getelementptr inbounds (i32, ptr addrspacecast (ptr addrspace(3) @twentyfive to ptr), i64 1), i64 0
  br i1 false, label %for.body65, label %for.end102

for.body65:
  call void @Imm_PS1_PT1_PS4_S7()
  %call69 = call i32 @foo(ptr addrspacecast (ptr addrspace(3) @twentythree to ptr), ptr addrspacecast (ptr addrspace(3) @twentyfour to ptr))
  ret i32 %call69

for.end102:
  %call106 = call i32 @Imm_PjPKjPS5_S6_b()
  ret i32 0
}

define void @Imm_PS1_PT1_PS4_S7() {
entry:
  store ptr addrspacecast (ptr addrspace(3) @twentysix to ptr), ptr null, align 8
  ret void
}

define i32 @Imm_PjPKjPS5_S6_b() {
entry:
  call void @Imm_PjPS4()
  ret i32 0
}

define void @Imm_PjPS4() {
entry:
  call void @Imm_PS1_Pj()
  ret void
}

define void @Imm_PS1_Pj() {
entry:
  store ptr addrspacecast (ptr addrspace(3) @twentyseven to ptr), ptr null, align 8
  ret void
}

define amdgpu_kernel void @EN64_compress_wrapperIhm() {
entry:
  %0 = call i32 @Ihm_one()
  ret void
}

define amdgpu_kernel void @EN64_compress_wrapperItm() {
entry:
  %0 = call i32 @Itm_one()
  ret void
}

define amdgpu_kernel void @EN64_compress_wrapperIjm() {
entry:
  %0 = call i32 @Ijm_one()
  ret void
}

define amdgpu_kernel void @EN64_compress_wrapperImm() {
entry:
  %0 = call i32 @Imm_one()
  ret void
}

```

JonChesterfield commented:
Alternative fix implemented at https://github.com/llvm/llvm-project/pull/161464
